operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/DefaultEventHandler.java

```diff
@@ -88,6 +90,12 @@ public void setEventSourceManager(DefaultEventSourceManager eventSourceManager)
 
   @Override
   public void handleEvent(Event event) {
+    // cache the latest version of the CR
```

Comment:
I honestly don't like this way of doing things. The behavior shouldn't depend on whether or not we have a cache or not. That's a bad design.

Reply:
Sure, but it doesn't change the fact that the behavior shouldn't depend on whether or not you optimize things by having a cache. A cache is an optimization, it's not a functional requirement and currently, it acts like one. This is the part that I don't like. The optimization should be removable and the code should still work when it is removed (slower, but still work). Right now, if you remove the "optimization" things don't work at all.

Reply:
"Optimizations should be removable." - Why is this a requirement? We do optimizations not because we want to remove them.

It's already there, the whole process it's much more efficient this way. Making a K8S API request on every event, could be quite an issue. So does not makes sense to change this because, Especially when the event comes from an event source. 
So feel free that makes it nicer / more readable, if you think this is not that nice, this way. But without breaking the functionality.

Reply:
Checking the operator sdk now, and this should work exactly the same. So the informer provides a cache about the objects.

Reply:
I'm not saying the optimization should be removed. I'm saying it should be removable without impacting the functionality, otherwise, it's not an optimization. Strictly speaking, there is no need for a cache, so the functionality should work even when there is no cache.

Reply:
Imho the best would be if you would create a separate issue, why is an un optimized version is needed. So we can discuss the pros and cons and all the arguments behind it.

Reply:
An unoptimized version is not needed, strictly speaking. But the algorithm should work without optimization. The current logic doesn't work without it. It's a smell, it means the logic is not sound. In the shared informer, the cache is transparent, the calling code doesn't know there's a cache. The same should apply here. If there's a cache, great, but the code shouldn't break if the cache is not there.

Again, there is nothing that says that you can't retrieve the resources on each call. If we want to prevent that, we should have a logic that acts like it's retrieving the resource on each call but returns a cached version when available instead of having the cache drive the core of the logic. Otherwise the logic is based on fragile assumptions that things are cached in a certain way and that will lead to things breaking in subtle ways. The logic shouldn't even care about a cache being there or not.

Reply:
I think I understand what you want to say:

> But the algorithm should work without optimization 

So I guess you are refereing to caching, so it should be more transparent (when asking for the lates custom resource). (This sentence can make sense in context of caches. Not in general - optimizes algorithms tends to be quite different from unoptimized ones).

So in this regard you are right, as I explained before, this was a simplest (quickest) approach that time to impement, and ment to be the first version, of this caching. 

I think the ideal would to contact the CustomResourceEventSource for the actual CustomResource - and not having the cache in this explicit way as it is now, it would be hidden behind the event source. 

In other words, as it is described here in caching section: https://csviri.medium.com/java-operator-sdk-introduction-to-event-sources-a1aab5af4b7b
The event source is responsible to make sure we have eventually  the fresh version of a resource. Currently this is true here, but in a not nice indirect way. So ideally where we are not reading the cache we would ask the eventsource for the latest resource, where the cr event source will decide if it takes it from local cache or calls the server. Eventually we can also replace the watches with informers there, so we delegate this caching to informer (but have to take a look deeper how fabric8 implements infomrers).

So in the cacing improvements issue I'm working on should reflect this. Would be better to discuss in detail in a call, if there are points which are not clear.

Just pls revert the changes in this PR (those ~4 lines), how the cache is propagted because it makes the implementation broken.

Reply:
OK, I will revert the change for now but we need to make the cache not part of the behavioral logic: the actions that need to be taken based on the state shouldn't depend on what is cached and what is not. Querying the event source for the Custom Resource seems reasonable to me, and, indeed, could return a cached version or not. I was thinking also about using informers as well and it would make sense to make sure that the client's implementation is working properly when it comes to caching.